### PR TITLE
Test operator restart when server pods roll

### PR DESCRIPTION
**Pull request by doxiao**
1) Add a new test case that restarts the operator pod in the middle of domain rolling restart.
2) Move the common code between the existing test classes and the new test class into common utils classes.
3) Clean up an unused parameter in isPodRestarted() method.

https://build.weblogick8s.org:8443/view/all/job/weblogic-kubernetes-operator-kind-new/504/

### CODE REVIEW

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+/**
+ * The common utility class for model-in-image tests.
+ */
+public class CommonMiiTestUtils {
+  /**
+   * Create a basic Kubernetes domain resource and wait until the domain is fully up.
+   *
+   */
+  public static void createMiiDomainAndVerify(
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java:35]: added

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    logger.info("Check admin service {0} is created in namespace {1}",
+        adminServerPodName, domainNamespace);
+    checkServiceExists(adminServerPodName, domainNamespace);
+
+    // check managed server services created
+    for (int i = 1; i <= replicaCount; i++) {
+      logger.info("Check managed server service {0} is created in namespace {1}",
+          managedServerPrefix + i, domainNamespace);
+      checkServiceExists(managedServerPrefix + i, domainNamespace);
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java:99]: This pattern is consistent in all test classes. Can you create a task to verify this and make the change to all classes if needed?

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorRestartWhenPodRoll.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+import static oracle.weblogic.kubernetes.utils.CommonPatchTestUtils.checkPodRestartVersionUpdated;
+import static oracle.weblogic.kubernetes.utils.CommonTestUtils.createSecretWithUsernamePassword;
+import static oracle.weblogic.kubernetes.utils.CommonTestUtils.installAndVerifyOperator;
+import static oracle.weblogic.kubernetes.utils.CommonTestUtils.verifyCredentials;
+import static org.awaitility.Awaitility.with;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+// Test to change the WebLogic credentials secret of a domain custom resource that uses model-in-image.
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorRestartWhenPodRoll.java]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorRestartWhenPodRoll.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+            opNamespace,
+            condition.getElapsedTimeInMS(),
+            condition.getRemainingTimeInMS()))
+        .until(assertDoesNotThrow(() -> operatorIsReady(opNamespace),
+          "operatorIsReady failed with ApiException"));
+
+    String opPodNameNew = 
+        assertDoesNotThrow(() -> getOperatorPodName(TestConstants.OPERATOR_RELEASE_NAME, opNamespace),
+        "Failed to get the name of the operator pod");
+
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/ItOperatorRestartWhenPodRoll.java:234]: added

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/TestActions.java [hunk 5]
```diff
@@ -910,6 +910,17 @@ public static V1PodList listPods(String namespace, String labelSelectors) throws
...
   }
 
+  /**
+   * Delete the operator pod.
+   *
+   * @param podName the name of the operator pod
+   * @param namespace the Kubernetes namespace that the operator belongs to
+   * @throws ApiException if Kubernetes client API call fails
+   */
+  public static void deleteOperatorPod(String podName, String namespace) throws ApiException {
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/actions/TestActions.java]: renamed

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java:35]: added

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  /**
+   * Create a domain object for a Kubernetes domain custom resource using the basic model-in-image image.
+   *
+   * @param domainResourceName name of the domain resource
+   * @param domNamespace Kubernetes namespace that the domain is hosted
+   * @param adminSecretName name of the new WebLogic admin credentials secret
+   * @param repoSecretName name of the secret for pulling the WebLogic image
+   * @param encryptionSecretName name of the secret for encryption
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java]: reworded

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonPatchTestUtils.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        "[\n  {\"op\": \"replace\", \"path\": \"/spec/%s\", \"value\": \"%s\"}\n]\n",
+        "webLogicCredentialsSecret/name", secretName);
+    logger.info("Patch the domain resource {0} in namespace {1} with: {2}\n",
+        domainResourceName, namespace, patch);
+
+    assertTrue(patchDomainCustomResource(
+        domainResourceName,
+        namespace,
+        new io.kubernetes.client.custom.V1Patch(patch),
+        io.kubernetes.client.custom.V1Patch.PATCH_FORMAT_JSON_PATCH),
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonPatchTestUtils.java]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonPatchTestUtils.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    return String.valueOf(newVersion);
+  }
+
+  /**
+   * Patch domain resource with a new WebLogic domain credentials secret and a new restartVersion,
+   * and verify if the domain spec has been correctly updated.
+   *
+   * @param domainUid name of the domain resource
+   * @param namespace Kubernetes namespace that the domain is hosted
+   * @param adminServerPodName name of the WebLogic admin server
```
[the remark this answers is not in the document]
Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonPatchTestUtils.java]: done

File: new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                .replicas(replicaCount)
+                .serverStartState("RUNNING"))
+            .configuration(new oracle.weblogic.domain.Configuration()
+                .model(new oracle.weblogic.domain.Model()
+                    .domainType("WLS")
+                    .runtimeEncryptionSecret(encryptionSecretName))
+                .introspectorJobActiveDeadlineSeconds(300L)));
+
+  }
+}
```
Comment (marinakog) [inline new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java:163]: can we add in this method image name param, so we can pass custom images as well?

Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java:166]: added

Comment (marinakog) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java:166]: could you please also have method createMiiDomainAndVerify which takes custom image name?

Comment (doxiao) [reply new-integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonMiiTestUtils.java:166]: done

**Pull request merged**